src/fields/field2_128/extend.rs

```diff
@@ -0,0 +1,355 @@
+    fft_array[index + recursive_len] += fft_array[index];
+}
+
+/// Implements procedure BUTTERFLY-BWD from Algorithm 1 in section 3.2 of [the paper][1].
+///
+/// [1]: https://eprint.iacr.org/2024/2010.pdf
+fn fft_butterfly_backward(
+    fft_array: &mut [Field2_128],
+    index: usize,
+    recursive_len: usize,
+    twiddle: Field2_128,
+) {
+    fft_array[index + recursive_len] -= fft_array[index];
+    fft_array[index] -= twiddle * fft_array[index + recursive_len];
+}
+
+/// Implements procedure BUTTERFLY-DIAG from Algorithm 1 in section 3.2 of [the paper][1].
+///
+/// [1]: https://eprint.iacr.org/2024/2010.pdf
+fn fft_butterfly_diagonal(
+    fft_array: &mut [Field2_128],
+    index: usize,
+    recursive_len: usize,
+    twiddle: Field2_128,
+) {
+    let prev_at_index = fft_array[index];
+
+    fft_array[index] -= twiddle * fft_array[index + recursive_len];
+    fft_array[index + recursive_len] += prev_at_index;
+}
+
+/// Direction in which the FFT operates.
+#[derive(Eq, PartialEq)]
+enum Direction {
+    Forward,
+    Backward,
+}
+
+/// Implements procedure FFT and IFFT (depending on direction) from Algorithm 1 in section 3.2
+/// of [the paper][1].
+///
+/// [1]: https://eprint.iacr.org/2024/2010.pdf
+fn fft(direction: Direction, power: u32, coset: usize, fft_array: &mut [Field2_128]) {
+    if power == 0 {
+        return;
+    }
+    let mut twiddled = vec![Field2_128::ZERO; 1 << (power - 1)];
+    for mut curr_power in 0..power {
+        // Forward FFT iterates over power..0
+        if direction == Direction::Forward {
+            curr_power = power - curr_power - 1;
+        }
+        let recursive_len = 1 << curr_power;
+
+        twiddles(curr_power, power, coset, &mut twiddled);
+
+        // for all u : 0 ≤ 2s · u < 2ℓ
+        for (index, start) in (0..1 << power).step_by(2 * recursive_len).enumerate() {
+            let twiddle = twiddled[index];
+            for v in 0..recursive_len {
+                match direction {
+                    Direction::Forward => {
+                        fft_butterfly_forward(fft_array, start + v, recursive_len, twiddle)
+                    }
+                    Direction::Backward => {
+                        fft_butterfly_backward(fft_array, start + v, recursive_len, twiddle)
+                    }
+                };
+            }
+        }
+    }
+}
+
+/// Perform a Fast Fourier Transform in the novel polynomial basis, in place.
+///
+/// The first `nodes_count` elements of `fft_array` are evaluations of a polynomial in one variable
+/// of degree up to `nodes_count - 1`. The FFT is used to interpolate and evaluate it.
+///
+/// On return, the first `nodes_count` elements of `fft_array` are coefficients of the polynomial
+/// and the remainder is evaluations of it at points `[nodes_count..fft_array.len()]`.
+///
+/// `fft_array.len()` must be `2^power` and greater than `nodes_count`. `coset` is which coset to
+/// recurse on and twiddle with.
```

Comment:
I think this is a clearer explanation of the coset parameter, and it gets away from the misleading bit manipulation view of "to twiddle".
```suggestion
/// `fft_array.len()` must be `2^power` and greater than `nodes_count`. `coset` selects the coset of
/// evaluation points at which to evaluate the polynomial.
```